### PR TITLE
fix: Make g.user attribute access safe for public users

### LINKED ISSUE #12618: How to enable query execution in sqllab for Anonymous user(public role) ?
Superset version: 0.38

Superset config:
AUTH_ROLE_PUBLIC='Public'
PUBLIC_ROLE_LIKE='Admin'

After superset init, all admin permissions are copied to Public role.

In Sqllab query execution does not work for Public role

ERROR: superset.views.core:Query2: 'AnonymousUserMixin' object has no attribute 'username'

Pls excuse me for image quality.

Query executes for admin user.

What need to be done to allow anonymous user to allow query execution in sqllab ?

### PR DESCRIPTION
**Pull request by robdiciuccio**
### SUMMARY
Make g.user attribute access safe for public users by:
1. Using the abstract `get_id` method
2. Checking for presence of `g.user.username` before using (this property is not set for public users)

### TEST PLAN
Automated tests should pass.

### ADDITIONAL INFORMATION

- [x] Has associated issue: fixes #12618
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Review by nytai — APPROVED** [review]

File: superset/db_engine_specs/base.py [hunk 1]
```diff
@@ -1024,7 +1024,7 @@ def estimate_query_cost(
         if not cls.get_allow_cost_estimate(extra):
             raise Exception("Database does not support cost estimation")
 
-        user_name = g.user.username if g.user else None
+        user_name = g.user.username if g.user and hasattr(g.user, "username") else None
```
Comment (dpgaspar) [inline superset/db_engine_specs/base.py:1027]: it probably makes more sense to just check if the user is authenticated, using `g.user.is_authenticated`

File: superset/charts/api.py [hunk 2]
```diff
@@ -954,7 +954,7 @@ def favorite_status(self, **kwargs: Any) -> Response:
         charts = ChartDAO.find_by_ids(requested_ids)
         if not charts:
             return self.response_404()
-        favorited_chart_ids = ChartDAO.favorited_ids(charts, g.user.id)
+        favorited_chart_ids = ChartDAO.favorited_ids(charts, g.user.get_id())
```
Comment (dpgaspar) [inline superset/charts/api.py:957]: for reference: https://github.com/maxcountryman/flask-login/blob/89ce0167a8603d2c53fa2d0d604a20bfe03c0eaf/flask_login/mixins.py#L75

**Review by dpgaspar — COMMENTED** [review]
nice! but we probably should check if the user is authenticated instead of checking if certain attrs are present, seems cleaner

File: superset/db_engine_specs/base.py (hunk 1 above)
Comment (robdiciuccio) [reply superset/db_engine_specs/base.py:1027]: I originally was checking `g.user.is_anonymous` in these use cases, but opted for the explicit check for username, as this is the value that is being fetched. There _may_ be use cases where username is not present in other states down the road, so thought explicit `hasattr` would be more future-proof.

Comment (dpgaspar) [reply superset/db_engine_specs/base.py:1027]: ok

**Review by dpgaspar — APPROVED** [review]

**Pull request merged**